README.md
<!-- -*- mode:gfm; word-wrap:nil; indent-tabs-mode:nil -*- github-flavored markdown -->

# OpenX Erlang Thrift Implementation

OX Thrift (`ox-thrift`) is a reimplementation of the
[Apache Thrift](https://thrift.apache.org/) Erlang library with an
emphasis on speed of encoding and decoding.  In its current
incarnation, it uses the structure definitions produced by the Apache
Thrift code generator.  However it has the following differences from
the Apache Thrift Erlang library.

* It supports only framed transports and binary and compact protocols.

* It gives up the ability to stream directly to or from the transport.
  Instead, the processor layer decodes the thrift message from a
  binary buffer and encodes to an iolist.  Compared to the Apache
  Erlang Thrift library, this simplifies the implementation and avoids
  a lot of record modifications that are inefficient in Erlang.

* The Thrift server uses the
  [Ranch](https://github.com/ninenines/ranch) acceptor pool instead of
  rolling its own.  OX Thrift requires ranch version 1.6 or newer.

* The `HandlerModule:handle_function(Function, Args)` interface
  expects the HandlerModule to take its arguments as a list instead of
  a tuple.  This is more consistent with Erlang conventions for
  functions that take a variable number of arguments.

* For the Thrift `map` type, for encoding OX Thrift will accept either
  a dict (as Apache Thrift does), a proplist, or an Erlang map.  For
  decoding, however, OX Thrift can be configured to return a dict or a
  map.  See the documentation for the `map_module` option.

* For the Thrift `set` type, for encoding OX Thrift will accept either a set
  (as Apache Thrift does) or a list.  OX Thrift does not deduplicate the list
  elements if passed a list.

* Like the Apache library, OX Thrift does not enforce required struct
  fields, on either encoding or decoding.

* Like the Apache library, OX Thrift does not populate the structs
  with default values on decoding.

## Interface

### Client

The OX Thrift client interface is provided by the `ox_thrift_client`
module.  This module exports two functions, `new`, and `call`.

#### `new` -- Manage a Connection to a Thrift Server

``` erlang
{ok, Client} = ox_thrift_client:new(Connection, ConnectionState, Transport, Protocol, Service)
{ok, Client} = ox_thrift_client:new(Connection, ConnectionState, Transport, Protocol, Service, Options)
```

* Connection: A module that manages a connection to the Thrift server.
  This module must support the interface defined by
  `ox_thrift_connection`.  The OX Thrift client will call the module's
  `checkout` and `checkin` functions.  The OX Thrift library supplies
  several modules that implement this interface, described in the
  [Connection Managers](#connection-managers) section below.
* ConnectionState: State managed by the Connection module.
* Transport: A module that provides the transport layer, such as
  `gen_tcp`.  This module is expected to export `send/2`, `recv/3`,
  and `close/1` functions.
* Protocol: A module that provides the Thrift protocol layer, e.g.,
  `ox_thrift_protocol_binary` or `ox_thrift_protocol_compact`
  if set to `undefined`, it will try to auto detect the protocol.
* Service: A module, produced by the Thrift IDL compiler from the
  service's Thrift definition, that provides the Service layer.
* Options: A list of options.
  * `{map_module, MapModule}` where `MapModule` is `dict` or `maps`
    specifies the module that is used to decode Thrift `map` fields.
    The default is `dict`.
  * `{recv_timeout, Milliseconds}` or `{recv_timeout, infinity}` The
    receive timeout.  The default is `infinity`.

The Connection module's `checkout` function should return a socket
Socket, and the OX Thrift library expects the Transport module to
support the following functions on this Socket.

* `ok = Transport:send(Socket, IOData)`,
* `{ok, Binary} = Transport:recv(Socket, Length, Timeout)`, and
* `ok = Transport:close(Socket)`

#### `call` -- Make a Call to a Thrift Server

``` erlang
{ok, OClient, Result} = ox_thrift_client:call(IClient, Function, Args)
```

* IClient: An `ox_thrift_client` record, as returned by the
  `ox_thrift_client:new` call or a previous `ox_thrift_client:call`
  call.
* Function: An atom representing the function to call.
* Args: A list containing the arguments to Function.

#### Connection Managers

OX Thrift supplies several connection managers for use with the OX
Thrift client.

* `ox_thrift_simple_socket` implements a non-reconnecting connection
  to the Thrift server.

* `ox_thrift_reconnecting_socket` implements a simple dedicated
  connection to the Thrift server that automatically reopens the
  connection when there is an error.

* `ox_thrift_socket_pool` implements a pool of connections that can be
  shared among multiple processes.  The pool is managed by a
  gen_server which you are expected to link into your supervisor tree
  by calling the `ox_thrift_socket_pool:start_link/4` function.
  See [the module](src/ox_thrift_socket_pool.erl) for more documentation.

* `ox_thrift_shackle_client` is an adapter to use OX Thrift with the
  [Shackle](https://github.com/lpgauth/shackle) socket pool.  See the
  [thrift_shackle_tests module](test/thrift_shackle_tests.erl) for an example
  of how to use it.

You can turn on some additional features out in `ox_thrift_socket_pool` by
adding the following to your `rebar.config` file:

``` erlang
{overrides,
  [ {add, ox_thrift,
      [ {erl_opts,
          [ {d, 'DEBUG_CONNECTIONS'} %% Debug errors in connection monitoring
          , {d, 'MONDEMAND_PROGID', myapp} %% Emit mondemand stats for errors.
          ]}
      ]}
  ]}.
```

* The `DEBUG_CONNECTIONS` macro controls whether the connections monitoring is
  logged if certain errors are detected.

* The `MONDEMAND_PROGID` macro controls whether error metrics are emitted
  using the [MonDemand](https://github.com/mondemand/mondemand-erlang) metric
  collection framework.

### Server

The OX Thrift server interface is provided by the `ox_thrift_server`
module.  This module is designed to be used with the
[Ranch](https://github.com/ninenines/ranch) acceptor pool.  Your
application needs to start Ranch, and then to register the
`ox_thrift_server` module with Ranch as a protocol handler.

```erlang
ranch:start_listener(
  ?THRIFT_SERVICE_REF,
  ranch_tcp,
  [ {port, Port}  ,  % https://github.com/ninenines/ranch/blob/master/doc/src/manual/ranch_tcp.asciidoc
    {num_acceptors, 10} ],
  ox_thrift_server,  % Ranch protocol module.
  #ox_thrift_config{
     service_module = service_thrift,
     protocol_module = ox_thrift_protocol_binary,
     handler_module = ?MODULE}).
```

Your Thrift server must supply two functions, `handle_function` and
`handle_error`.  These functions are defined by the `ox_thrift_server`
behaviour, and the module that supplies these functions is specified
in the `handler_module` field.

Unlike the Apache Thrift's `handle_function` Erlang interface, OX
Thrift passes the Args parameter as a list instead of a tuple.

```erlang
-behaviour(ox_thrift_server).

handle_function (Function, Args) when is_atom(Function), is_list(Args) ->
  case apply(?MODULE, Function, Args) of
    ok    -> ok;
    Reply -> {reply, Reply}
  end.

handle_error(Function, Reason) ->
  ok.
```

The `handle_function` function may return `ok` for a void function, or
`{reply, Reply}` for a non-void function.  In addition, the function
may request that the Thrift server close the client's connection after
returning the reply by returning `{ok, close}` for a void function or
`{reply, Reply, close}` for a non-void function.

If the Thrift function wants to return one of its declared exceptions,
`E`, it may call `{throw, E}`.  The Thrift server will catch the
exception and return a message to the client, where the exception will
be re-thrown.

The protocol is specified by setting the `protocol_module` field to
one of the following values.

* `ox_thrift_protocol_binary`: binary protocol
* `ox_thrift_protocol_compact`: compact protocol

#### Thrift `map` Field Decoding

By default, Thrift `map` fields are decoded to Erlang `dict`s.  You
can specify that they be decoded as Erlang `maps` with the
`map_module` option.

``` erlang
#ox_thrift_config{options = [ { map_module, maps } ]}
```

#### Setting the Server's `recv` Timeout

The default `recv` timeout is `infinity`, which means that the server
will keep a socket open indefinitely waiting for a client to send a
request.  You can override this with the `recv_timeout` option.

``` erlang
#ox_thrift_config{options = [ { recv_timeout, TimeoutMilliseconds } ]}
```

#### Limiting the Maximum Message Size

By default the OX Thrift server will accept Thrift messages of unlimited size
(`infinity`).  If you are in an environment where clients may sent you
improperly-framed thrift messages (for example, random requests coming from a
port scanner), ox_thrift allows you to limit the maximum size of the message
the server will accept.  Messages that are larger than this limit will result
in a call to the `handle_error` function with a `large_message` error.

``` erlang

ChangeLog
* Wed 18 Jan 2023 Kenan Gillet <kenan.gillet@openx.com> 0.22.0
- ox_thrift_server: detect protocol module can now be turned off
  by setting #ox_thrift_config.protocol_module to a specific module

* Wed 25 May 2022 Kenan Gillet <kenan.gillet@openx.com> 0.21.0
- remove support for erlang:get_stacktrace

* Mon Jan 6 2020 David Hull <david.hull@openx.com> 0.20.0
- Improvements to ox_thrift_socket_pool:
  * Move socket open from gen_server to client so that it does not block other
    requests.
  * Add checkout and open stats.
  * Add spares_factor option to maintain idle pool.
  * Add a simple DNS cache.
- Add ox_thrift_shackle_client module.

* Thu May 23 2019 David Hull <david.hull@openx.com> 0.19.0
- ox_thrift_socket_pool: Split unavailable stat into error_pool_full and
  error_connect.
- ox_thrift_socket_pool: Add close_local, close_remote, and close_die stats.

* Tue May 21 2019 David Hull <david.hull@openx.com> 0.18.0
- Add a max_age_jitter_ms option to ox_thrift_socket_pool to add some
  randomness to max_age_ms with a default of 20%.

* Mon Apr 29 2019 David Hull <david.hull@openx.com> 0.17.0
- Update for ranch 1.7.1.  Ranch >= 1.6.0 is now required.

* Tue Nov 6 2018 David Hull <david.hull@openx.com> 0.16.1
- Fix for ox_thrift_client gen_tcp:recv enomem error when response is more
  than 64 MB.

* Mon Oct 15 2018 David Hull <david.hull@openx.com> 0.16.0
- Convert reply to binary when spawning ox_thrift_server handle_function call.

* Fri Oct 5 2018 Kenan Gillet <kenan.gillet@openx.com> 0.15.0
- Add ox_thrift_server spawn_options option, which causes a new process to be
  spawned to handle each Thrift function call.  At OpenX this gives another 4%
  improvement in encoding and decoding since version 0.14.1 when it is used.

* Fri Sep 28 2018 David Hull <david.hull@openx.com> 0.14.1
- Further optimize encoding.  This change gives about a 25% cumulative
  improvement in encoding time since version 0.13.1.

* Wed Sep 26 2018 David Hull <david.hull@openx.com> 0.14.0
- Optimize decoding and encoding of struct fields.  These changes give about a
  15% improvement in decoding and encoding times since version 0.13.1.

* Thu Sep 13 2018 David Hull <david.hull@openx.com> 0.13.1
- Inline list functions in protocol modules.
- Replace a lists:map with a list comprehension in protocol modules.
- Avoid an extra list when encoding collections in protocol modules.

* Mon Aug 27 2018 David Hull <david.hull@openx.com> 0.13.0
- Fix skipping of unknown structs when decoding.

* Tue Aug 21 2018 David Hull <david.hull@openx.com> 0.12.0
- Improve specs and add some guards to ox_thrift_socket_pool.
- Add check for debugging a crash in ox_thrift_socket_pool.
- Drop framewerk build support.

* Thu Jun 21 2018 David Hull <david.hull@openx.com> 0.11.1
- Decode and encode framed transport message length as unsigned.

* Thu May 17 2018 David Hull <david.hull@openx.com> 0.11.0
- Add a max_message_size option to ox_thrift_server.  Requests larger than
  this are treated as an error.  (They are probably actually non-thrift
  payloads sent to the thrift socket.)

* Thu May 24 2018 David Hull <david.hull@openx.com> 0.10.5
- Transform a throw of an unexpected type from the Thrift server's
  handle_function to an "unknown" application exception.

* Thu May 24 2018 David Hull <david.hull@openx.com> 0.10.4
- Fix declaration of {map_module, maps} in ox_thrift_option type.

* Fri Mar 16 2018 David Hull <david.hull@openx.com> 0.10.3
- Fix closed socket error handling in ox_thrift_client.
- Close socket in ox_thrift_reconnecting_socket to prevent resource
  leakage.

* Fri Feb 9 2018 David Hull <david.hull@openx.com> 0.10.2
- Define NAMESPACED_TYPES for Erlang versions 18 and above.

* Tue Dec 5 2017 David Hull <david.hull@openx.com> 0.10.1
- Fix bug in ServiceModule:function_info call for oneway functions
  introduced in version 0.9.0.

* Mon Nov 27 2017 Kenan Gillet <kenan.gillet@openx.com> 0.10.0
- Expose pool statistics for ox_thrift_socket_pool

* Fri Sep 29 2017 David Hull <david.hull@openx.com> 0.9.0
- Add the map_module option to allow user to specify whether map
  fields should be returned as an Erlang dict or a map.

* Thu Mar 23 2017 David Hull <david.hull@openx.com> 0.8.0
- Rework thrift client connection handling.
- The thrift client now returns an error tuple to indicate an error
  instead of throwing an error.
- When encoding a map type, accept an Erlang map.
- Auto-detect call protocol (binary or compact) in ox_thrift_server.

* Thu Nov 10 2016 David Hull <david.hull@openx.com> 0.7.0
- APPS-14103: Add compact protocol support.


src/ox_thrift_protocol_compact.erl
%% Copyright 2016-2018, OpenX.  All rights reserved.
%% Licensed under the conditions specified in the accompanying LICENSE file.

-module(ox_thrift_protocol_compact).

%% For a description of the compact protocol, see
%% https://erikvanoosten.github.io/thrift-missing-specification/.

%% For a description of the zigzag and base 128 integer encoding for protobuf,
%% see https://developers.google.com/protocol-buffers/docs/encoding.

-include("ox_thrift_internal.hrl").

-export([ encode_call/4, encode_message/5, decode_message/3 ]).

-export([ encode_record/2, decode_record/2, decode_record/3 ]).

-define(THRIFT_PROTOCOL, compact).
-include("ox_thrift_protocol.hrl").

-compile({inline, [ write_message_begin/3
                  , write_field/5
                  , write_field_stop/1
                  , write_map_begin/4
                  , write_list_or_set_begin/3
                  , write/3
                  , write_varint/1
                  , read_message_begin/1
                  , read_field_begin/2
                  , read_map_begin/1
                  , read_list_or_set_begin/1
                  , read/2
                  , read_varint/1
                  , encode_zigzag/1
                  , decode_zigzag/1
                  ]}).
-compile(inline_list_funcs).

-define(TYPE_STRUCT_FIELD_STOP, 0).
-define(TYPE_STRUCT_FIELD_TRUE, 1).
-define(TYPE_STRUCT_FIELD_FALSE, 2).
-define(TYPE_STRUCT_FIELD_BYTE, 3).
-define(TYPE_STRUCT_FIELD_I16, 4).
-define(TYPE_STRUCT_FIELD_I32, 5).
-define(TYPE_STRUCT_FIELD_I64, 6).
-define(TYPE_STRUCT_FIELD_DOUBLE, 7).
-define(TYPE_STRUCT_FIELD_STRING, 8).
-define(TYPE_STRUCT_FIELD_LIST, 9).
-define(TYPE_STRUCT_FIELD_SET, 10).
-define(TYPE_STRUCT_FIELD_MAP, 11).
-define(TYPE_STRUCT_FIELD_STRUCT, 12).

term_to_wire_struct(bool)       -> ?TYPE_STRUCT_FIELD_TRUE; % list/set/map
term_to_wire_struct(true)       -> ?TYPE_STRUCT_FIELD_TRUE; % struct field
term_to_wire_struct(false)      -> ?TYPE_STRUCT_FIELD_FALSE;
term_to_wire_struct(byte)       -> ?TYPE_STRUCT_FIELD_BYTE;
term_to_wire_struct(double)     -> ?TYPE_STRUCT_FIELD_DOUBLE;
term_to_wire_struct(i16)        -> ?TYPE_STRUCT_FIELD_I16;
term_to_wire_struct(i32)        -> ?TYPE_STRUCT_FIELD_I32;
term_to_wire_struct(i64)        -> ?TYPE_STRUCT_FIELD_I64;
term_to_wire_struct(string)     -> ?TYPE_STRUCT_FIELD_STRING;
term_to_wire_struct({struct, _})-> ?TYPE_STRUCT_FIELD_STRUCT;
term_to_wire_struct({map, _, _})-> ?TYPE_STRUCT_FIELD_MAP;
term_to_wire_struct({set, _})   -> ?TYPE_STRUCT_FIELD_SET;
term_to_wire_struct({list, _})  -> ?TYPE_STRUCT_FIELD_LIST.

wire_to_term_struct(?TYPE_STRUCT_FIELD_TRUE)    -> true;
wire_to_term_struct(?TYPE_STRUCT_FIELD_FALSE)   -> false;
wire_to_term_struct(?TYPE_STRUCT_FIELD_BYTE)    -> byte;
wire_to_term_struct(?TYPE_STRUCT_FIELD_DOUBLE)  -> double;
wire_to_term_struct(?TYPE_STRUCT_FIELD_I16)     -> i16;
wire_to_term_struct(?TYPE_STRUCT_FIELD_I32)     -> i32;
wire_to_term_struct(?TYPE_STRUCT_FIELD_I64)     -> i64;
wire_to_term_struct(?TYPE_STRUCT_FIELD_STRING)  -> string;
wire_to_term_struct(?TYPE_STRUCT_FIELD_STRUCT)  -> struct;
wire_to_term_struct(?TYPE_STRUCT_FIELD_MAP)     -> map;
wire_to_term_struct(?TYPE_STRUCT_FIELD_SET)     -> set;
wire_to_term_struct(?TYPE_STRUCT_FIELD_LIST)    -> list.

wire_to_term_lsm(?TYPE_STRUCT_FIELD_TRUE)    -> bool;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_BYTE)    -> byte;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_DOUBLE)  -> double;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_I16)     -> i16;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_I32)     -> i32;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_I64)     -> i64;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_STRING)  -> string;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_STRUCT)  -> struct;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_MAP)     -> map;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_SET)     -> set;
wire_to_term_lsm(?TYPE_STRUCT_FIELD_LIST)    -> list.

%% -define(DEBUG_READ, true).

%% There is an asymmetry to the `write/1' and `read/2' interfaces.  `write/1'
%% returns an iolist for the converted data, which is expected to be
%% integrated into a larger iolist for the entire message, while `read/2'
%% parses the data from a binary buffer and returns the parsed data and the
%% remaining buffer.

-define(COMPACT_BINARY_HEADER, 16#82).
-define(COMPACT_BINARY_VERSION, 1).

write_message_begin (Name, Type, SeqId) ->
  %% Write a version 1 message header.
  TypeAndVersion = Type * 32 + ?COMPACT_BINARY_VERSION,
  [ ?COMPACT_BINARY_HEADER, TypeAndVersion, write_varint(SeqId), write(string, Name, []) ].

write_field (Type, Id, LastId, Data, Acc) ->
  DeltaId = Id - LastId,
  if DeltaId > 0 andalso DeltaId < 16 ->
      case Type of
        bool -> case Data of
                  true  -> [ DeltaId * 16 + ?TYPE_STRUCT_FIELD_TRUE | Acc ];
                  false -> [ DeltaId * 16 + ?TYPE_STRUCT_FIELD_FALSE | Acc ]
                end;
        byte            -> [ DeltaId * 16 + ?TYPE_STRUCT_FIELD_BYTE, Data | Acc ];
        _               -> [ DeltaId * 16 + term_to_wire_struct(Type) | encode(Type, Data, Acc) ]
      end;
     true ->
      case Type of
        bool -> case Data of
                  true  -> [ ?TYPE_STRUCT_FIELD_TRUE, write_varint(encode_zigzag(Id)) | Acc ];
                  false -> [ ?TYPE_STRUCT_FIELD_FALSE, write_varint(encode_zigzag(Id)) | Acc ]
                end;
        byte            -> [ ?TYPE_STRUCT_FIELD_BYTE, write_varint(encode_zigzag(Id)), Data | Acc ];
        _               -> [ term_to_wire_struct(Type), write_varint(encode_zigzag(Id)) | encode(Type, Data, Acc) ]
      end
  end.

write_field_stop (Acc) ->
  [ ?TYPE_STRUCT_FIELD_STOP | Acc ].

write_map_begin (KType, VType, Size, Acc) ->
  case Size of
    0 -> [ 0 | Acc ];
    _ -> KTypeWire = term_to_wire_struct(KType),
         VTypeWire = term_to_wire_struct(VType),
         [ write_varint(Size), KTypeWire * 16 + VTypeWire | Acc ]
  end.

write_list_or_set_begin (EType, Size, Acc) ->
  ETypeWire = term_to_wire_struct(EType),
  if Size < 15 -> [ Size * 16 + ETypeWire | Acc ];
     true      -> [ 16#F0 + ETypeWire, write_varint(Size) | Acc ]
  end.


write (bool, true, Acc) ->
  [ 2 | Acc ];

write (bool, false, Acc) ->
  [ 0 | Acc ];

write (byte, Byte, Acc) ->
  [ <<Byte:8/signed>> | Acc ];

write (i16, I16, Acc) ->
  [ write_varint(encode_zigzag(I16)) | Acc ];

write (i32, I32, Acc) ->
  [ write_varint(encode_zigzag(I32)) | Acc ];

write (i64, I64, Acc) ->
  [ write_varint(encode_zigzag(I64)) | Acc ];

write (double, Double, Acc) ->
  [ <<Double:64/float-signed-little>> | Acc ];

write (string, Str, Acc) when is_list(Str) ->
  Bin = list_to_binary(Str),
  BinLen = size(Bin),
  [ write_varint(BinLen), Bin | Acc ];

write (string, Bin, Acc) when is_binary(Bin) ->
  BinLen = size(Bin),
  [ write_varint(BinLen), Bin | Acc ].


write_varint (UnsignedInt) when UnsignedInt < 128 ->
  [ UnsignedInt ];
write_varint (UnsignedInt) ->
  [ 128 + UnsignedInt rem 128 | write_varint(UnsignedInt div 128) ].


-ifdef(DEBUG_READ).
read(DataIn, Type) ->
  io:format("read(~p, ~p) ->\n", [ DataIn, Type ]),
  Ret = {DataOut, Val} = read_real(DataIn, Type),
  io:format("  {~p, ~p}\n", [ DataOut, Val ]),
  Ret.
-define(READ, read_real).
-else. %% ! DEBUG_READ
-define(READ, read).
-endif. %% ! DEBUG_READ


-spec read_message_begin(IData::binary()) -> {OData::binary(), Name::binary(), Type::integer(), SeqId::integer()}.
read_message_begin (Data0) ->
  case Data0 of
    <<?COMPACT_BINARY_HEADER, Type:3, ?COMPACT_BINARY_VERSION:5, Data1/binary>> ->
      {Data2, SeqId} = read_varint(Data1),
      {Data3, Name} = read(Data2, string),
      {Data3, Name, Type, SeqId};
    _  ->
      %% Unexpected header or version number.
      error({bad_header, binary_part(Data0, {0, 2})})
  end.

-spec read_field_begin (IData::binary(), LastId::integer())
                       -> {OData::binary(), field_stop}
                        | {OData::binary(), bool, Id::integer(), Val::boolean()}
                        | {OData::binary(), byte, Id::integer(), Val::integer()}
                        | {OData::binary(), i16 | i32 | i64 | double | struct | map | set | list, Id::integer()}.
read_field_begin (Data0, LastId) ->
  %% This function decodes and returns the value when it's "easy" (bool,
  %% byte), and leaves it to the caller when it's hard.
  case Data0 of
    <<?TYPE_STRUCT_FIELD_STOP:8, Data1/binary>> ->
      {Data1, field_stop};
    %% long-form header fields (DeltaId is 0, Id is encoded after type field):
    <<0:4, ?TYPE_STRUCT_FIELD_TRUE:4, Data1/binary>> ->
      {Data2, IdZ} = read_varint(Data1),
      {Data2, bool, decode_zigzag(IdZ), true};
    <<0:4, ?TYPE_STRUCT_FIELD_FALSE:4, Data1/binary>> ->
      {Data2, IdZ} = read_varint(Data1),
      {Data2, bool, decode_zigzag(IdZ), true};
    <<0:4, ?TYPE_STRUCT_FIELD_BYTE:4, Data1/binary>> ->
      {Data2, IdZ} = read_varint(Data1),
      <<Val:8/signed, Data3/binary>> = Data2,
      {Data3, byte, decode_zigzag(IdZ), Val};
    <<0:4, Type:4, Data1/binary>> ->
      {Data2, IdZ} = read_varint(Data1),
      {Data2, wire_to_term_struct(Type), decode_zigzag(IdZ)};
    %% short-form header fields:
    <<DeltaId:4, ?TYPE_STRUCT_FIELD_TRUE:4, Data1/binary>> ->
      {Data1, bool, DeltaId + LastId, true};
    <<DeltaId:4, ?TYPE_STRUCT_FIELD_FALSE:4, Data1/binary>> ->
      {Data1, bool, DeltaId + LastId, false};
    <<DeltaId:4, ?TYPE_STRUCT_FIELD_BYTE:4, Val:8/signed, Data1/binary>> ->
      {Data1, byte, DeltaId + LastId, Val};
    <<DeltaId:4, Type:4, Data1/binary>> ->
      {Data1, wire_to_term_struct(Type), DeltaId + LastId}
  end.

%% This isn't necessary, since we never explicitly read a `field_stop', we
%% just find it when trying to read a `field_begin'.
%%
%% ?READ (field_stop, Data0) ->
%%   {?tType_STOP, Data1} = read(?tType_BYTE, Data0),
%%   {ok, Data1};

-spec read_map_begin(IData::binary())
                    -> {OData::binary(), 'undefined', 'undefined', 0}
                     | {OData::binary(), KType::proto_type(), VType::proto_type(), Size::pos_integer()}.
read_map_begin (Data0) ->
  {Data1, Size} = read_varint(Data0),
  if Size =:= 0 -> {Data1, undefined, undefined, Size};
     true       -> <<KType:4, VType:4, Data2/binary>> = Data1,
                   {Data2, wire_to_term_lsm(KType), wire_to_term_lsm(VType), Size}
  end.

-spec read_list_or_set_begin(IData::binary()) -> {OData::binary(), EType::proto_type(), Size::integer()}.
read_list_or_set_begin (Data0) ->
  <<CompactSize:4, EType:4, Data1/binary>> = Data0,
  case CompactSize of
    16#F -> {Data2, Size} = read_varint(Data1),
            {Data2, wire_to_term_lsm(EType), Size};
    _    -> {Data1, wire_to_term_lsm(EType), CompactSize}
  end.


?READ (Data0, bool) ->
  <<Bool:8, Data1/binary>> = Data0,
  {Data1, Bool =/= 0};

?READ (Data0, byte) ->
  <<Val:8/signed, Data1/binary>> = Data0,
  {Data1, Val};

?READ (Data0, i16) ->
  {Data1, Z} = read_varint(Data0),
  {Data1, decode_zigzag(Z)};

?READ (Data0, i32) ->
  {Data1, Z} = read_varint(Data0),
  {Data1, decode_zigzag(Z)};

?READ (Data0, i64) ->
  {Data1, Z} = read_varint(Data0),
  {Data1, decode_zigzag(Z)};

?READ (Data0, double) ->
  <<Val:64/float-signed-little, Data1/binary>> = Data0,
  {Data1, Val};

?READ (Data0, string) ->
  {Data1, Size} = read_varint(Data0),
  <<String:Size/binary, Data2/binary>> = Data1,
  {Data2, String}.


read_varint (Data) -> read_varint(Data, 0, 1).

read_varint (Data0, Acc, Multiplier) ->
  <<Val:8/integer-unsigned, Data1/binary>> = Data0,
  if Val < 128 -> {Data1, Val * Multiplier + Acc};
     true      -> read_varint(Data1, (Val - 128) * Multiplier + Acc, Multiplier * 128)
  end.


%% -spec read_data(DataIn::binary(), Size::non_neg_integer()) -> {DataOut::binary(), Value::binary()}.
%% read_data (Data0, Size) ->
%%   case Size of
%%     0 -> {Data0, <<>>};
%%     _ -> <<Val:Size/binary, Data1/binary>> = Data0,
%%          {Data1, Val}
%%   end.


encode_zigzag (SignedInt) ->
  if SignedInt >= 0 -> SignedInt * 2;
     true           -> -SignedInt * 2 - 1
    end.


decode_zigzag (ZigZagInt) ->
  if ZigZagInt rem 2 =:= 1 -> - (ZigZagInt + 1) div 2;
     true                  -> ZigZagInt div 2
  end.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").
-endif.

-ifdef(EUNIT).

write_varint_test () ->
  ?assertEqual(<<0>>, iolist_to_binary(write_varint(0))),
  ?assertEqual(<<1>>, iolist_to_binary(write_varint(1))),
  ?assertEqual(<<127>>, iolist_to_binary(write_varint(127))),
  ?assertEqual(<<128, 1>>, iolist_to_binary(write_varint(128))),
  ?assertEqual(<<185, 96>>, iolist_to_binary(write_varint(12345))),
  ?assertEqual(<<135, 173, 75>>, iolist_to_binary(write_varint(1234567))),
  ?assertEqual(<<255, 127>>, iolist_to_binary(write_varint(16383))),
  ?assertEqual(<<128, 128, 1>>, iolist_to_binary(write_varint(16384))),
  ?assertEqual(<<255, 255, 127>>, iolist_to_binary(write_varint(2097151))),
  ?assertEqual(<<255, 255, 255, 127>>, iolist_to_binary(write_varint(268435455))).

read_varint_test () ->
  ?assertEqual({<<>>, 0}, read_varint(<<0>>)),
  ?assertEqual({<<>>, 1}, read_varint(<<1>>)),
  ?assertEqual({<<>>, 127}, read_varint(<<127>>)),
  ?assertEqual({<<>>, 128}, read_varint(<<128, 1>>)).


varint_roundtrip_test () ->
  lists:map(
    fun (I) -> ?assertMatch({{<<>>, I}, _, I},
                            begin
                              B = iolist_to_binary(write_varint(I)),
                              {read_varint(B), B, I}
                            end) end,
    [ 0, 1, 9, 10, 99, 100, 9999, 16#10001000, 16#FFFFFFFF, 16#FEDCBA9876543210 ]).


encode_zigzag_test () ->
  ?assertEqual(0, encode_zigzag(0)),
  ?assertEqual(2, encode_zigzag(1)),
  ?assertEqual(1, encode_zigzag(-1)),
  ?assertEqual(254, encode_zigzag(127)),
  ?assertEqual(255, encode_zigzag(-128)),
  ?assertEqual(4294967294, encode_zigzag(2147483647)),
  ?assertEqual(4294967295, encode_zigzag(-2147483648)).

zigzag_roundtrip_test () ->
  lists:map(
    fun (I) -> ?assertMatch({I, _, I},
                            begin
                              Z = encode_zigzag(I),
                              {decode_zigzag(Z), Z, I}
                            end) end,
    [ 0, 1, 2, 3, -1, -2, -3, 100, -100, 999, -999, 16#7FF0, 99999999, -99999999 ]).


message_test () ->
  Name = <<"test">>,
  NameLen = size(Name),
  Type = ?tMessageType_CALL,
  SeqId = 16#7FFFFFF0,
  ?assertEqual({<<>>, Name, Type, SeqId},
               read_message_begin(iolist_to_binary(write_message_begin(Name, Type, SeqId)))),

  SeqIdEnc = iolist_to_binary(write_varint(SeqId)),
  ?assertEqual({<<>>, Name, Type, SeqId},
               read_message_begin(<<?COMPACT_BINARY_HEADER, Type:3, ?COMPACT_BINARY_VERSION:5, SeqIdEnc/binary, NameLen, Name/binary>>)).

field_test () ->
  ?assertEqual(<<17>>, iolist_to_binary(write_field(bool, 1, 0, true, []))),
  ?assertEqual(<<18>>, iolist_to_binary(write_field(bool, 1, 0, false, []))),
  ?assertEqual(<<1, 2>>, iolist_to_binary(write_field(bool, 1, 99, true, []))),
  ?assertEqual(<<2, 2>>, iolist_to_binary(write_field(bool, 1, 99, false, []))),
  ?assertEqual(<<19, 123>>, iolist_to_binary(write_field(byte, 1, 0, 123, []))),
  ?assertEqual(<<3, 2, 123>>, iolist_to_binary(write_field(byte, 1, 99, 123, []))),
  ?assertEqual(<<16#25, 16#54>>, iolist_to_binary(write_field(i32, 32, 30, 42, []))),
  ?assertEqual(<<16#05, 64, 16#54>>, iolist_to_binary(write_field(i32, 32, 0, 42, []))),

  %% Type = i32,
  %% Id = 16#7FF0,
  %% Name is not sent in binary protocol.
  %% ?assertEqual({<<>>, Type, Id}, read_field_begin(iolist_to_binary(write_field_begin(Type, Id, 0)), 0)),
  %% ?assertEqual({<<>>, Type, Id}, read_field_begin(iolist_to_binary(write_field_begin(Type, Id, Id - 5)), Id - 5)),
  ok.

map_test () ->
  KType = byte,
  VType = string,
  Size0 = 16#7FFFFFF1, %% Normal map header.
  ?assertEqual({<<>>, KType, VType, Size0}, read_map_begin(iolist_to_binary(write_map_begin(KType, VType, Size0, [])))),
  Size1 = 0, %% Empty map header.
  ?assertEqual({<<>>, undefined, undefined, Size1}, read_map_begin(iolist_to_binary(write_map_begin(KType, VType, Size1, [])))).

list_or_set_test () ->
  EType = byte, %% Large set header.
  Size0 = 16#7FFFFFF2,
  ?assertEqual({<<>>, EType, Size0}, read_list_or_set_begin(iolist_to_binary(write_list_or_set_begin(EType, Size0, [])))),
  Size1 = 14, %% Compact set header.
  ?assertEqual({<<>>, EType, Size1}, read_list_or_set_begin(iolist_to_binary(write_list_or_set_begin(EType, Size1, [])))).

basic_test () ->
  B = 16#7F,
  ?assertEqual({<<>>, B}, read(iolist_to_binary(write(byte, B, [])), byte)),

  S = 16#7FFF,
  ?assertEqual({<<>>, S}, read(iolist_to_binary(write(i16, S, [])), i16)),

  I = 16#7FFFFFFF,
  ?assertEqual({<<>>, I}, read(iolist_to_binary(write(i32, I, [])), i32)),

  F = 1234.25,
  ?assertEqual({<<>>, F}, read(iolist_to_binary(write(double, F, [])), double)),

  SB = <<"hello, world">>,
  ?assertEqual({<<>>, SB}, read(iolist_to_binary(write(string, SB, [])), string)),

  SL = "hello, world",
  ?assertEqual({<<>>, SB}, read(iolist_to_binary(write(string, SL, [])), string)).

-endif. %% EUNIT
